### DISCUSSION TITLE
Defining behavior in a hierarchy, inherting behavior from parent and making children visible

### DISCUSSION THREAD

**Discussion by Caligino** [bevyengine/bevy#5880] [Q&A] [answered]
So what i'm trying to do is get my camera and light source to rotate around the 3D model to give an illusion of motion to the object since i couldnt get the logo to rotate around its own center. Maybe i'm making this too complicated since i already wrote code that compiles? i might as well keep going in this direction. If there _is_ a solution for the original approach i'm all ears. but if someone could tell me what im doing wrong with this approach thatd be great too.

`use bevy::{prelude::*};`

`use std::f32::consts::TAU;`

`#[derive(Component)]`
`struct Rotatable {`
 `   speed: f32,`
`}`
`fn main() {`
  `  App::new()`
 `       .insert_resource(ClearColor(Color::rgb(0.2,0.2, 0.2)))`
 `       .insert_non_send_resource(WindowDescriptor {`
`            width: 1020.0,`
 `           height: 720.0,`
  `         title: "logo".to_string(),`
    `        resizable: false,`
  `          ..Default::default() })`
 `     //  .add_startup_system(GTransform)`
  `      .add_startup_system(setup)`
   `     .add_system(rotate_obj)`
`       // .add_startup_system(spawn_glb)`
  `     // .add_startup_system(spawn_lighting)`
        
 `       .add_plugins(DefaultPlugins)`
 `       .run();`
`}`

`//fn rotator_system(time: Res<Time>, mut query: Query<&mut Transform, `With<Rotatable>>) {`
` //   for mut transform in &mut query {`
`  //      transform.rotate_x(3.0 * time.delta_seconds());`
`   // }`
`//}`

`fn setup(mut commands: Commands,  ass: Res<AssetServer>) {`

`   // .insert(Rotatable { speed: 0.;3})`
`   //let position = global_transform.translation(); `

`let _parent = commands.spawn_bundle(Camera3dBundle{`
 `   transform: Transform::from_xyz(0.0,1.0,1.0).looking_at( Vec3::new(0.0,0.3,0.0), `Vec3::Y),`
 `   ..default()`

   
`})`
`.insert(Rotatable { speed: 0.3})`
`.with_children(|parent|{`
``
``
`parent.spawn_bundle(PointLightBundle{`
`    transform: Transform::from_xyz(0.5,1.0,1.0),`
 ``  
  ` ..default()`
  ``
 `      })`
  ``      .insert(Rotatable { speed: 0.3});``
  ``   
  ``
``    
 ` //spawn glb object`
  ``     
`  let logo = ass.load("logo.glb#Scene0");`
`  let xf =  Transform::from_xyz(0.0,1.0,-1.0)`
`  .with_rotation(Quat::from_rotation_x(1.5 ));`
 `` 
` parent.spawn_bundle(SceneBundle {`
 ` scene: logo, `
`  transform: xf,`
 `..default()`
`});`
``
`});`
`}`
``
``
``
`fn rotate_obj(mut objects: Query<(&mut Transform, &Rotatable)>, timer: `Res<Time>) {`
`    for (mut transform, object) in &mut objects {`
`        // The speed is first multiplied by TAU which is a full rotation (360deg) in `radians,`
`        // and then multiplied by delta_seconds which is the time that passed last `frame.`
 `    ``  // In other words. Speed is equal to the amount of rotations per second.`
  `    ``  transform.rotate_y( object.speed * std::f32::consts::TAU * `timer.delta_seconds());`
  `         }`
    `    }`

_note:_ sorry the formatting sucks. i hate this implementation why cant they just use html?

**Comment by wilk10** [comment, marked as answer]
To fix the formatting I think you need to wrap everything in:
` ```rust `
your code
` ``` `

Regarding your question, for the hierarchy to work properly, the parent and all children need to have all components from the `SpatialBundle` (meaning `Transform`, `GlobalTransform`, `Visibility`, `ComputedVisibility`). The `Camera3dBundle` doesn't have `Visibility` and `ComputedVisibility`, so you need to add those manually to the entity with that component (together they form the `VisibilityBundle`, to insert both in one go). I _think_ this should fix it.

**Reply by Caligino** [reply]
ok i can hardly find any documentation on spatialbundle, but i have been somewhat aware thats what i need. thank you for clarifying that.

**Reply by Sheepyhead** [reply]
It's mentioned in the 0.8 release notes, at least

**Reply by nicopap** [reply]
There is a PR open to print a warning in the console if we detect a hierarchy with a missing link https://github.com/bevyengine/bevy/pull/5590
